File: 01-programming/csharp/concepts/code/01-basics/simple-arithmetic.cs
// -o-o-o-o- Simple Arithmetic -o-o-o-o-

/* This file goes over basic arithmetic operations in C# simply.
 *
 *  # # # # KEY # # # #
 *
 * +  - addition operator
 * -  - subtraction operator
 * *  - multiplication operator
 * /  - division operator
 *    Note: dividing two integers gives an integer result (remainder dropped).
 *          eg. 7 / 2 = 3 (not 3.5). Cast to double for decimal result.
 * %  - modulus operator. Returns the remainder of a division.
 *    eg. 7 % 2 = 1
 *
 * Console.WriteLine() - prints text to the console with a newline at the end.
 *
 * $"..."  - string interpolation. Lets you embed variables directly in a string.
 *           eg. $"Result: {a + b}"
 *           Equivalent to printf("Result: %i", a + b) in C.
 *
 * (double) - casting. Converts an integer to a double for decimal division.
 *            eg. (double)a / b
 *            Same concept as casting in C.
 *
 * */

using System;

class Program {
    static void Main() {

        int a = 10;
        int b = 3;

        // Basic arithmetic
        Console.WriteLine($"{a} + {b} = {a + b}");   // Addition
        Console.WriteLine($"{a} - {b} = {a - b}");   // Subtraction
        Console.WriteLine($"{a} * {b} = {a * b}");   // Multiplication
        Console.WriteLine($"{a} / {b} = {a / b}");   // Integer division (remainder dropped)
        Console.WriteLine($"{a} % {b} = {a % b}");   // Modulus (remainder)

        // Casting to get a decimal result from division
        double decimalResult = (double)a / b;
        Console.WriteLine($"\n{a} / {b} as decimal = {decimalResult}");

        // Compound assignment operators (shorthand)
        int x = 10;
        x += 5;  // same as x = x + 5
        Console.WriteLine($"\nx after += 5: {x}");
        x -= 3;  // same as x = x - 3
        Console.WriteLine($"x after -= 3: {x}");
        x *= 2;  // same as x = x * 2
        Console.WriteLine($"x after *= 2: {x}");
        x /= 4;  // same as x = x / 4
        Console.WriteLine($"x after /= 4: {x}");

    }
}

File: 01-programming/python/concepts/code/01-basics/simple-arithmetic.py
# -o-o-o-o- Simple Arithmetic -o-o-o-o-

# This file goes over basic arithmetic operations in Python simply.
#
#  # # # # KEY # # # #
#
# +   - addition operator. Same as C and C#.
# -   - subtraction operator. Same as C and C#.
# *   - multiplication operator. Same as C and C#.
# /   - division operator. ALWAYS returns a float in Python, even for integers.
#       eg. 7 / 2 = 3.5   (NOT 3 like in C and C#)
#       This is different from C and C# where int / int = int.
# //  - integer (floor) division. Drops the decimal, like int / int in C and C#.
#       eg. 7 // 2 = 3
# %   - modulus operator. Returns the remainder of a division. Same as C and C#.
#       eg. 7 % 2 = 1
# **  - exponentiation (power) operator. Unique to Python.
#       eg. 2 ** 3 = 8  (2 to the power of 3)
#       No direct operator equivalent in C/C# (use pow() or Math.Pow()).
#
# print()  - prints text to the console with a newline at the end.
# f"..."   - f-string. Embeds variables directly in a string.
#            Equivalent to $"..." in C# or printf() in C.
#
# Note: Python integers have unlimited size — no overflow like in C or C#.
#

a = 10
b = 3

# Basic arithmetic
print(f"{a} + {b}  = {a + b}")    # Addition
print(f"{a} - {b}  = {a - b}")    # Subtraction
print(f"{a} * {b}  = {a * b}")    # Multiplication
print(f"{a} / {b}  = {a / b}")    # True division (always float)
print(f"{a} // {b} = {a // b}")   # Integer / floor division
print(f"{a} % {b}  = {a % b}")    # Modulus (remainder)
print(f"{a} ** {b} = {a ** b}")   # Exponentiation (10 to the power of 3)

# Division always returns a float in Python
print(f"\n6 / 2 = {6 / 2}")       # 3.0, not 3
print(f"6 // 2 = {6 // 2}")      # 3 (integer division)

# Compound assignment operators (shorthand)
x = 10
x += 5   # same as x = x + 5
print(f"\nx after += 5:  {x}")
x -= 3   # same as x = x - 3
print(f"x after -= 3:  {x}")
x *= 2   # same as x = x * 2
print(f"x after *= 2:  {x}")
x //= 4  # same as x = x // 4
print(f"x after //= 4: {x}")
x **= 2  # same as x = x ** 2
print(f"x after **= 2: {x}")

# Python integers have no overflow limit
big_number = 2 ** 100
print(f"\n2 ** 100 = {big_number}")
